compiler/src/dotty/tools/dotc/transform/ElimRepeated.scala

```diff
@@ -34,132 +34,209 @@ class ElimRepeated extends MiniPhase with InfoTransformer { thisPhase =>
+   *          which forwards java varargs to `ddef`. It retains all the
+   *          flags of `ddef` except `Private`.
    *
-   *  A bridge is necessary because the following hold
+   *  A bridge is necessary because the following hold:
    *    - the varargs in `ddef` will change from `RepeatedParam[T]` to `Seq[T]` after this phase
-   *    - _but_ the callers of `ddef` expect its varargs to be changed to `Array[? <: T]`, since it overrides
-   *      a Java varargs
+   *    - _but_ the callers of `ddef` expect its varargs to be changed to `Array[? <: T]`
    *  The solution is to add a "bridge" method that converts its argument from `Array[? <: T]` to `Seq[T]` and
    *  forwards it to `ddef`.
    */
-  private def addVarArgsBridge(ddef: DefDef)(using Context): Tree = {
+  private def addVarArgsBridge(ddef: DefDef, javaOverride: Boolean)(using ctx: Context): Tree =
     val original = ddef.symbol.asTerm
+    // For simplicity we always set the varargs flag
+    // although it's not strictly necessary for overrides
+    // (but it is for non-overrides)
+    val flags = ddef.symbol.flags | JavaVarargs
+
+    // The java-compatible bridge symbol
     val bridge = original.copy(
-      flags = ddef.symbol.flags &~ Private | Artifact,
-      info = toJavaVarArgs(ddef.symbol.info)).enteredAfter(thisPhase).asTerm
-    val bridgeDef = polyDefDef(bridge, trefs => vrefss => {
-      val (vrefs :+ varArgRef) :: vrefss1 = vrefss
-      // Can't call `.argTypes` here because the underlying array type is of the
-      // form `Array[? <: SomeType]`, so we need `.argInfos` to get the `TypeBounds`.
-      val elemtp = varArgRef.tpe.widen.argInfos.head
-      ref(original.termRef)
-        .appliedToTypes(trefs)
-        .appliedToArgs(vrefs :+ tpd.wrapArray(varArgRef, elemtp))
-        .appliedToArgss(vrefss1)
-    })
-
-    Thicket(ddef, bridgeDef)
-  }
+        // non-overrides cannot be synthetic otherwise javac refuses to call them
+        flags = if javaOverride then flags | Artifact else flags,
+        info = toJavaVarArgs(ddef.symbol.info)
+      ).asTerm
+
+    currentClass.info.member(bridge.name).alternatives.find { s =>
+      s.matches(bridge) &&
+      !(javaOverride && s.asSymDenotation.is(JavaDefined))
+    } match
+      case Some(conflict) =>
+        ctx.error(s"@varargs produces a forwarder method that conflicts with ${conflict.showDcl}", original.sourcePos)
+        ddef
+      case None =>
+        val bridgeDef = polyDefDef(bridge.enteredAfter(thisPhase), trefs => vrefss => {
+          val init :+ (last :+ vararg) = vrefss
+          // Can't call `.argTypes` here because the underlying array type is of the
+          // form `Array[? <: SomeType]`, so we need `.argInfos` to get the `TypeBounds`.
+          val elemtp = vararg.tpe.widen.argInfos.head
+
+          // The generation of the forwarding call needs to be deferred, otherwise
+          // generic and curried methods won't pass the tree checker.
+          atNextPhase {
```

Comment:
I realize `atNextPhase` already existed in the compiler but I would rather not use it: it's meaning is relative to the phase in the current context, which in this method happens to always be `thisPhase` because it's called from `transformDefDef` which runs `atPhase(thisPhase)` (because it needs to see repeated param types before they're eliminated). Using `atPhase(thisPhase.next)` instead is clearer. (though I think it might be even better to not change the phase back and forth like this and only use `atPhase(thisPhase)` in the specific places where we need to look at repeated param types).

Reply:
I've gone the "better" way you've offered: no more `atNextPhase`, only `atPhase(thisPhase)` where necessary